versions/models.py

```diff
@@ -1071,15 +1076,22 @@ def clone_relations(self, clone, manager_field_name):
             destination.add(item)
 
         # retrieve all current m2m relations pointing the newly created clone
-        m2m_rels = source.through.objects.filter(**{source.source_field.attname: clone.id})  # filter for source_id
+        # filter for source_id
+        m2m_rels = list(source.through.objects.filter(**{source.source_field.attname: clone.id}))
+        later = []
         for rel in m2m_rels:
             # Only clone the relationship, if it is the current one; Simply adjust the older ones to point the old entry
             # Otherwise, the number of pointers pointing an entry will grow exponentially
             if rel.is_current:
-                rel.clone(forced_version_date=self.version_end_date)
+                later.append(rel.clone(forced_version_date=self.version_end_date, in_bulk=True))
             # On rel, set the source ID to self.id
             setattr(rel, source.source_field_name, self)
-            rel.save()
+            if not hasattr(rel, '_not_created') or not rel._not_created:
+                rel.save()
+        # Perform the bulk changes rel.clone() did not perform because of the in_bulk parameter
+        # This saves a huge bunch of SQL queries
+        source.through.objects.filter(id__in=[l.id for l in later]).update(**{'version_start_date': forced_version_date})
+        source.through.objects.bulk_create([r for r in m2m_rels if hasattr(r, '_not_created') and r._not_created])
```

Comment:
Hm, I'll write such a test, no problem...
However, I don't see any duplicate entries in your code - there are 3 insert or update operations:
1.) `save()` is called on those relations not having the `_not_created` property and being in the `m2m_rels`-list => these are typically the non-current entries still pointing the current model-object
2.) `update()` is called on objects inside the `later`-list. `later` only contains clones of current objects. But due to the properties of a CleanerVersion-current object (`version_end_date` is `None`, `version_start_date` indicates the beginning of the version's validity, `version_birth_date` indicates the birth date of the object and `id` (the unique ID)` == identity` (the object ID)), we only want to update the `version_start_date` on new 'current' objects.
3.) `bulk_create()` is called in order to create all the versions that were current prior to calling `clone()`, on line 1086. These versions have their `version_end_date` set (thanks to code in `def clone(...)`) and `id` has been assigned a new value, such that `id != identity` (lines 1028 & 1029).

In a nutshell, I don't think `save()` (or any other inserting command) is called twice on a same object, provoking duplicate entries.
However, better safe than sorry - I'll write the tests in order to be 100% sure. ;)